### DISCUSSION TITLE
How to traine model on PyTorch Lightning + Huggingface

### DISCUSSION THREAD

**Discussion by basecl** [community/community#52228] [Programming Help]
### Select Topic Area

Question

### Body

Can you please send me a code example on how to make a fine-tune model such as Stable Diffusion v1.5 or another text to image model on PyTorch Lightning using Huggingface?

I just couldn't find an example of this code, and it would be cool if someone could throw.

**Comment by siennaphia** [comment]
Hiii this is a pretty basic code example on how to make a fine tune model on PyTorch Lightning ussing Huggingface. For this to work you need to replace train_dataloader and val_dataloader with your own data. 

`from transformers import BertForSequenceClassification, AdamW, get_linear_schedule_with_warmup
from pytorch_lightning import LightningModule, Trainer
import torch
class BertClassifier(LightningModule):
    def __init__(self, model_name='bert-base-uncased', num_labels=2, learning_rate=2e-5):
        super().__init__()
        self.save_hyperparameters()
        self.bert = BertForSequenceClassification.from_pretrained(self.hparams.model_name, num_labels=self.hparams.num_labels)
        self.learning_rate = self.hparams.learning_rate

    def forward(self, input_ids, attention_mask, labels=None):
        output = self.bert(input_ids, attention_mask=attention_mask, labels=labels)
        return output.loss, output.logits

    def training_step(self, batch, batch_idx):
        input_ids = batch['input_ids']
        attention_mask = batch['attention_mask']
        labels = batch['labels']
        loss, logits = self(input_ids, attention_mask, labels)
        self.log('train_loss', loss, prog_bar=True)
        return loss

    def validation_step(self, batch, batch_idx):
        input_ids = batch['input_ids']
        attention_mask = batch['attention_mask']
        labels = batch['labels']
        loss, logits = self(input_ids, attention_mask, labels)
        self.log('val_loss', loss, prog_bar=True)
        return loss

    def test_step(self, batch, batch_idx):
        input_ids = batch['input_ids']
        attention_mask = batch['attention_mask']
        labels = batch['labels']
        loss, logits = self(input_ids, attention_mask, labels)
        self.log('test_loss', loss, prog_bar=True)
        return loss

    def configure_optimizers(self):
        optimizer = AdamW(self.parameters(), lr=self.learning_rate)
        scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=0, num_training_steps=self.trainer.max_steps)
        return [optimizer], [scheduler]

model = BertClassifier()
trainer = Trainer(max_epochs=3, gpus=1)
trainer.fit(model, train_dataloader, val_dataloader)

`

A LightningModule organizes your PyTorch code. You can think of it as a standardized way to build a PyTorch model. It contains everything related to your model such as the model's forward method, the training loop, the validation loop, optimizers, learning rate schedulers etc.
In the defined class BertClassifier, we first initialize the pre-trained BERT model for sequence classification, and set the learning rate. Depending on your specific use case, you might need to customize the BertClassifier class and the training process but this iss a general overview. 

The forward function defines the forward pass for our model. It takes the inputs to the model, feeds them through the model and returns the loss and logits. The training_step method defines what happens during training. It takes a batch of data, feeds it to the model, and logs the training loss. Similarly, the validation_step and test_step methods define what happens during validation and testing respectively.
The configure_optimizers method defines the optimizers and the learning rate schedulers.

The training of the model is handled by PyTorch Lightning's Trainer. The trainer takes the model and the data, and handles the training process.This basic pattern of using PyTorch Lightning can be adapted and evolved to fit almost any training regime, and can be used with any model from Huggingface's transformers library, not just BERT.

Here are the sources I used!
https://lightning.ai/docs/pytorch/stable/notebooks/lightning_examples/text-transformers.html
https://huggingface.co/transformers/v3.3.1/training.html
https://stackoverflow.com/questions/69249187/how-to-test-a-model-before-fine-tuning-in-pytorch-lightning

**Reply by basecl** [reply]
It's good code, but I meant it a little differently. In the Stable Diffusion model for example, the input is a photo and a text and how to train it ?

**Comment by morpheuslord** [comment]
I have the same kind of doubt I want to train the Llama2 model with a CVE dataset and also with a malware detection dataset but I am not able to find ample resources for this task.

**Comment by Awrsha** [comment]
Here is an example code for fine-tuning a text-to-image model like Stable Diffusion in PyTorch Lightning:

```python
import pytorch_lightning as pl  
from torchvision.transforms import Normalize  
from stable_diffusion import StableDiffusion

class StableDiffusionFineTuner(pl.LightningModule):
    
    def __init__(self):
      
        self.model = StableDiffusion.from_pretrained("stablediffusion/stable-diffusion-v1_5-ansi125").to("cuda")
        
        self.loss = torch.nn.MSELoss()
        
    def forward(self, input_text):
        return self.model.generate_images(input_text)
        
    def training_step(self, batch, batch_idx):
        
        images, texts = batch 
        
        # Normalize images
        images = Normalize([0.5, 0.5, 0.5],  [0.5, 0.5, 0.5])(images)
            
        # Forward pass
        outputs = self.forward(texts)  
        
        loss = self.loss(outputs, images)
        return loss
        
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-5)    
        return optimizer
```

This runs the stable diffusion model in `forward()`, calculates the MSE loss between the generated image and ground truth image in `training_step()`, and uses Adam optimizer with a learning rate of 1e-5 in `configure_optimizers()`.

Hope this helps! Let me know if you have any other questions.

**Reply by basecl** [reply]
It doesn't work, can you reset the whole code?

**Reply by RamtinKosari** [reply]
Hi @bomjtawerok 
Below is an example code snippet that demonstrates how to fine-tune a text-to-image model, specifically the Stable Diffusion v1.5 model, using PyTorch Lightning and the Huggingface library. Please note that this is a simplified example, and you may need to modify it based on your specific requirements and the dataset you're working with :
```python
import torch
import pytorch_lightning as pl
from torchvision.transforms import transforms
from PIL import Image

from transformers import CLIPProcessor, CLIPModel, CLIPTextToImageGenerator

# Define your dataset and dataloader here

class TextToImageFineTuner(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.clip_processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
        self.clip_model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
        self.generator = CLIPTextToImageGenerator.from_pretrained("openai/clip-vit-base-patch32")

    def forward(self, inputs):
        return self.generator.generate(**inputs)

    def training_step(self, batch, batch_idx):
        texts, images = batch
        inputs = self.clip_processor(texts=texts, images=images, return_tensors="pt", padding=True)

        # Generate images from text
        generated_images = self(inputs)

        # Calculate loss (you may need to define your own loss function)
        loss = ...

        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-4)
        return optimizer

# Initialize the fine-tuner and trainer

fine_tuner = TextToImageFineTuner()
trainer = pl.Trainer(gpus=1, max_epochs=10)  # Adjust the parameters as per your requirements

# Start fine-tuning

trainer.fit(fine_tuner, train_dataloader)  # Provide your training dataloader

# Generate images from text

text_inputs = ["A cat sitting on a mat", "A beautiful sunset on the beach"]
image_inputs = [Image.open("cat.jpg"), Image.open("sunset.jpg")]  # Provide your own images

inputs = fine_tuner.clip_processor(texts=text_inputs, images=image_inputs, return_tensors="pt", padding=True)
generated_images = fine_tuner(inputs)

# Save or display the generated images

for i, image in enumerate(generated_images):
    image.save(f"generated_image_{i}.jpg")
```
In this example, we use the CLIP (Contrastive Language-Image Pretraining) model provided by Huggingface. First, we initialize the necessary components such as the CLIP processor, CLIP model, and the CLIP text-to-image generator. Then, we define the training step, configure the optimizer, and initialize the fine-tuner and trainer.

To fine-tune the model, you need to provide your own dataset and dataloader. Additionally, you might need to define your own loss function based on the specific task you're working on.

After fine-tuning, you can generate images from text by providing text inputs and corresponding image inputs to the fine-tuner. The generated images can then be saved or displayed as per your requirements.

**Reply by morpheuslord** [reply]
This is an example of text-to-image but I cannot use this to train a text generation model I have prepared a dataset https://github.com/morpheuslord/CVE-llm_dataset but I have had zero success training it I tried everything like using autotrain, from scratch training code and also used runpod.io for the job but nothing worked.

**Comment by morpheuslord** [comment]
Hello, I managed to implement a fine tune for the `meta-llama/Llama-2-7b-hf` model with my custom dataset using this code:
```python
import torch
from datasets import load_dataset
from peft import LoraConfig
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, TrainingArguments
from trl import SFTTrainer

def finetune_llama_v2():
    data = load_dataset("morpheuslord/cve-llm-training", split="train")
    tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")
    tokenizer.pad_token = tokenizer.eos_token
    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype="float16", bnb_4bit_use_double_quant=True
    )
    
    # Load the model on the CPU and move it to GPU later during training
    model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", quantization_config=bnb_config)
    model.config.use_cache = False
    model.config.pretraining_tp = 1
    
    peft_config = LoraConfig(
        r=64, lora_alpha=32, lora_dropout=0.05, bias="none", task_type="CAUSAL_LM"
    )
    training_arguments = TrainingArguments(
        output_dir="secllama",
        per_device_train_batch_size=1,  # Reduce batch size even further
        gradient_accumulation_steps=16,  # Further increase gradient accumulation
        optim="paged_adamw_8bit",
        learning_rate=2e-4,
        lr_scheduler_type="linear",
        save_strategy="epoch",
        logging_steps=30,
        num_train_epochs=10,
        max_steps=30,
        fp16=True,
        push_to_hub=True
    )
    
    # Move the model to GPU inside the trainer
    trainer = SFTTrainer(
        model=model,
        train_dataset=data,
        peft_config=peft_config,
        dataset_text_field="output",
        args=training_arguments,
        tokenizer=tokenizer,
        packing=False,
        max_seq_length=256  # Further reduce sequence length
    )
    trainer.train()
    trainer.push_to_hub()

if __name__ == "__main__":
    finetune_llama_v2()
```

And the final updated code looks like this 
![image](https://github.com/community/community/assets/70637311/2b882de2-d6c9-482c-bf87-1491122ecd4c)

But when I try to download and implement it for testing it's showing. `config.json` missing and 
```bash
OSError: morpheuslord/secllama does not appear to have a file named pytorch_model.bin, tf_model.h5, model.ckpt 
or flax_model.msgpack.
```

What can I do to make this work and is my code wrong?

**Reply by andysingal** [reply]
following

**Reply by morpheuslord** [reply]
??. I wanted to ask how can I export my model to have the missing files or use the existing files or the outputs in my case to run the model 😅
